util.go

```diff
@@ -0,0 +1,80 @@
+package files
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"sort"
+)
+
+// ToFile is an alias for n.(File). If the file isn't a regular file, nil value
+// will be returned
+func ToFile(n Node) File {
+	f, _ := n.(File)
+	return f
+}
+
+// ToDir is an alias for n.(Directory). If the file isn't directory, a nil value
+// will be returned
+func ToDir(n Node) Directory {
+	d, _ := n.(Directory)
+	return d
+}
+
+// FileFrom is a convenience function which tries to extract or create new file
+// from provided value. If a passed value can't be turned into a File, nil will
+// be returned.
+//
+// Supported types:
+// * files.File (cast from Node)
+// * DirEntry / DirIterator (cast from e.Node())
+// * []byte (wrapped into NewReaderFile)
+// * io.Reader / io.ReadCloser (wrapped into NewReaderFile)
+func FileFrom(n interface{}) File {
+	switch f := n.(type) {
+	case File:
+		return f
+	case DirEntry:
+		return ToFile(f.Node())
+	case []byte:
+		return NewReaderFile(ioutil.NopCloser(bytes.NewReader(f)), nil)
+	case io.ReadCloser:
+		return NewReaderFile(f, nil)
+	case io.Reader:
+		return NewReaderFile(ioutil.NopCloser(f), nil)
+	default:
+		return nil
+	}
+}
+
+// DirFrom is a convenience function which tries to extract or create new
+// directory from the provided value. If a passed value can't be turned into a
+// Directory, nil will be returned.
+//
+// Supported types:
+// * files.File (cast from Node)
+// * DirEntry (cast from e.Node())
+// * DirIterator (current file, cast from e.Node())
+// * []DirEntry (wrapped into NewSliceFile)
+// * map[string]Node (wrapped into NewSliceFile)
+func DirFrom(n interface{}) Directory {
+	switch f := n.(type) {
+	case Directory:
+		return f
+	case DirEntry:
+		return ToDir(f.Node())
+	case []DirEntry:
+		return NewSliceFile(f)
```

Comment:
Do these *have* to be sorted? If this is going to be an API, we should probably be handling that either here or in the `NewSliceFile` constructor.

(the `NewSliceFile` (or `NewSliceDirectory`) constructor will probably also need to be able to return an error in case of duplicates)

Basically, if we're going to *expose* this, we need to be extra careful about checking assumptions (users are *going* to pass unsorted directories).

Reply:
> Do these have to be sorted?

Not really, iterators don't guarantee that entries will be sorted (because we want to be able to stream when listing from unixfs / `ls --stream`).

Reply:
Ah, got it (just saw the sort below and thought that was a requirement).